packages/blockchain-link/src/workers/baseWebsocket.ts

```diff
@@ -1,115 +1,27 @@
-    connectionTimeout?: number;
-    keepAlive?: boolean;
-    agent?: WebSocket.ClientOptions['agent'];
-    headers?: WebSocket.ClientOptions['headers'];
-    onSending?: (message: Record<string, any>) => void;
-}
-
-const DEFAULT_TIMEOUT = 20 * 1000;
-const DEFAULT_PING_TIMEOUT = 50 * 1000;
-
-type EventMap = { [event: string]: any };
-
-type WsEvents = {
-    error: string;
-    disconnected: undefined;
-};
-
-export abstract class BaseWebsocket<T extends EventMap> extends TypedEmitter<T & WsEvents> {
-    readonly options: Options;
-
-    private readonly messages;
+export abstract class BaseWebsocket<T extends Record<string, any>> extends WebsocketClient<T> {
     private readonly subscriptions: Subscription<keyof T>[] = [];
-    private readonly emitter: TypedEmitter<WsEvents> = this;
-
-    private ws?: WebSocket;
-    private pingTimeout?: TimerId;
-    private connectPromise?: Promise<void>;
-
-    protected abstract ping(): Promise<unknown>;
-
-    protected abstract createWebsocket(): WebSocket;
-
-    constructor(options: Options) {
-        super();
-        this.options = options;
-        this.messages = createDeferredManager({
-            timeout: this.options.timeout || DEFAULT_TIMEOUT,
-            onTimeout: this.onTimeout.bind(this),
-        });
-    }
-
-    private setPingTimeout() {
-        if (this.pingTimeout) {
-            clearTimeout(this.pingTimeout);
-        }
-        this.pingTimeout = setTimeout(
-            this.onPing.bind(this),
-            this.options.pingTimeout || DEFAULT_PING_TIMEOUT,
-        );
-    }
 
-    private onTimeout() {
-        const { ws } = this;
-        if (!ws) return;
-        this.messages.rejectAll(new CustomError('websocket_timeout'));
-        ws.close();
-    }
-
-    private async onPing() {
+    async onPing() {
         // make sure that connection is alive if there are subscriptions
-        if (this.ws && this.isConnected()) {
-            try {
-                if (this.subscriptions.length > 0 || this.options.keepAlive) {
-                    await this.ping();
-                } else {
-                    this.ws.close();
-                }
-            } catch {
-                // empty
-            }
+        if (this.subscriptions.length > 0 || this.options.keepAlive) {
+            await this.ping();
+        } else {
+            this.disconnect();
         }
     }
 
-    private onError() {
-        this.onClose();
-    }
-
-    protected sendMessage(message: Record<string, any>) {
-        const { ws } = this;
-        if (!ws) throw new CustomError('websocket_not_initialized');
-        const { promiseId, promise } = this.messages.create();
-
-        const req = { id: promiseId.toString(), ...message };
-
-        this.setPingTimeout();
-
-        this.options.onSending?.(message);
-
-        ws.send(JSON.stringify(req));
-
-        return promise;
-    }
-
-    protected onMessage(message: string) {
+    protected onMessage(message: WebsocketResponse) {
```

Comment:
Probably not very important, but these overriden `onMessage` don't call `setPingTimeout` as the base one does, so the pings will be timed differently.